storage/backend-test-suite/src/basic.rs

```diff
@@ -87,9 +87,65 @@ fn put_twice_then_commit_read_last<B: Backend>(backend: B) {
     assert_eq!(dbtx.get(IDX.0, b"hello"), Ok(Some(b"b".as_ref())),);
 }
 
+fn put_iterator_count_matches<B: Backend>(backend: B) {
+    let store = backend.open(desc(1)).expect("db open to succeed");
+
+    let mut dbtx = store.transaction_rw();
+    dbtx.put(IDX.0, vec![0x00], vec![]).unwrap();
+    dbtx.put(IDX.0, vec![0x01], vec![]).unwrap();
+    dbtx.put(IDX.0, vec![0x02], vec![]).unwrap();
+    dbtx.put(IDX.0, vec![0x03], vec![]).unwrap();
+    dbtx.commit().expect("commit to succeed");
+
+    let dbtx = store.transaction_ro();
+    assert_eq!(dbtx.prefix_iter(IDX.0, vec![]).unwrap().count(), 4);
+}
+
+fn put_and_iterate_over_prefixes<B: Backend>(backend: B) {
+    let store = backend.open(desc(1)).expect("db open to succeed");
+
+    // Populate the database with some
+    let mut dbtx = store.transaction_rw();
+    dbtx.put(IDX.0, b"ac".to_vec(), b"2".to_vec()).unwrap();
+    dbtx.put(IDX.0, b"bf".to_vec(), b"7".to_vec()).unwrap();
+    dbtx.put(IDX.0, b"ab".to_vec(), b"1".to_vec()).unwrap();
+    dbtx.put(IDX.0, b"aca".to_vec(), b"3".to_vec()).unwrap();
+    dbtx.put(IDX.0, b"bz".to_vec(), b"8".to_vec()).unwrap();
+    dbtx.put(IDX.0, b"x".to_vec(), b"9".to_vec()).unwrap();
+    dbtx.put(IDX.0, b"bb".to_vec(), b"6".to_vec()).unwrap();
+    dbtx.put(IDX.0, b"b".to_vec(), b"5".to_vec()).unwrap();
+    dbtx.put(IDX.0, b"acb".to_vec(), b"4".to_vec()).unwrap();
+    dbtx.put(IDX.0, b"aa".to_vec(), b"0".to_vec()).unwrap();
+    dbtx.commit().expect("commit to succeed");
+
+    // Check for a non-existent prefix
+    assert_eq!(
+        store.transaction_ro().prefix_iter(IDX.0, b"foo".to_vec()).unwrap().next(),
+        None,
+    );
+
+    // Check for items that are supposed to be present
+    let check = |range: std::ops::RangeInclusive<usize>, prefix: Data| {
+        let dbtx = store.transaction_ro();
+        let vals = dbtx.prefix_iter(IDX.0, prefix).unwrap().map(|(_key, val)| val);
+        let expected_vals = range.map(|x| Data::from(x.to_string()));
+        assert!(vals.eq(expected_vals));
+        drop(dbtx);
+    };
+
+    check(0..=9, b"".to_vec());
+    check(0..=4, b"a".to_vec());
+    check(0..=0, b"aa".to_vec());
+    check(2..=4, b"ac".to_vec());
+    check(5..=8, b"b".to_vec());
+    check(9..=9, b"x".to_vec());
+}
+
```

Comment:
Can you please add a test of iteration that starts with a set, then deletes some items of the set arbitrarily in the middle, and then iterates again and ensures only the correct items are left?

Reply:
Btw, I meant within and outside of a transaction. 

This can be seen in multiple cases:
1. Create the data, commit, erase a subset, check, commit, check
2. Create the data, erase a subset, check, commit, check

Reply:
> Btw, I meant within and outside of a transaction.
> 
> This can be seen in multiple cases:
> 
> 1. Create the data, commit, erase a subset, check, commit, check
> 2. Create the data, erase a subset, check, commit, check

These are covered by randomised property tests but will add some normal tests as well since these tend to be easier to analyse.

Reply:
Added some more tests for deleting stuff